### ISSUE TITLE
Safari version <17.4 support (Promise.withResolvers from pdf.js 4.5)

### ISSUE THREAD

**Issue by DaanPepping** [stephanrauh/ngx-extended-pdf-viewer#2536]
**I need your help!**
I've been using your library for a while and it has saved me tons of hours so first of i'd like to thank you for your hard work on maintaining this library!

Recently i've upgraded my capacitor application to angular 18 and i've been able to upgrade ngx-extended-pdf-viewer to 21.3.5. All seemed good at first but when i started testing on iOS and MacOS i ran into the dreaded Promise.withResolvers is undefined error:
<img width="375" alt="image" src="https://github.com/user-attachments/assets/14dfbdda-c9c2-4ae5-8aed-b8e2710ee04c">

I did some digging and i can confirm that the issue is caused by pdf.js >4.5, since running older versions work fine. 
I tried doing the polyfill workaround mentioned in other issue threads but no dice.

Some more digging confirmed that safari v17.4 added Promise.withResolvers support which explains why it doesn't work on my end since i'm running safari v17.1. The easy fix would be to require safari v17.4 or higher to use my application but this version was released earlier this year and would significantly bump my required iOS version which i would rather avoid.

in previous versions the forceUsingLegacyES5 property was available which from my understanding loads an older version of pdf.js , but this was moved to the PDFScriptLoaderService in version 21. 
Unfortunately my restrictive CSP requirements are locking me out of using version 20.5.x which is the lowest version with angular 18 support.  So i'm at a bit of a crossroads, as to how i should proceed.

Would it be possible to reintroduce the forceUsingLegacyES5 property so i could run an older version of pdf.js 
or should i bite the bullet early and restrict access from "legacy" iOS users? 

EDIT:
I have found a hacky workaround by using version 20.2.0 but manually modify my package.lock.json file and bump the  peerDependencies on @angular/common and @angular/core from <18 to <19. Surprisingly this seems to work in my case but it feels like a very bad practice.

**Comment by stephanrauh** [comment]
Let's start with your hack. It's sort of OK to use it because I simply forgot to declare Angular 18 compatibility. You can make you hack a bit less hacky by adding some lines to the `package.json`. I had to do this in the showcase because another library also forgot to update to Angular 18:

<img width="761" alt="image" src="https://github.com/user-attachments/assets/dbaa2bc0-3c3d-4b63-afcc-30a988a8da85">

**Comment by stephanrauh** [comment]
If I remember correctly, using the ES5 build won't help you because it also contains `Promise.withResolvers`. So, yes, I want to bring the attribute `forceUsingLegacyES5` back, but I have more homework to do.

In the meantime, can you debug my Polyfill code? I'd like to know what's going on, so I can fix the bug properly, but I can't downgrade to Safari 17.1. Here's the line that should add the Polyfill:
https://github.com/stephanrauh/ngx-extended-pdf-viewer/blob/09f6728d563482b6dd50101fb8382dc991fc7dc9/projects/ngx-extended-pdf-viewer/src/lib/pdf-script-loader.service.ts#L275

At the moment, I suspect that the viewer detects that it needs the ES5 version, and that this leads to skipping the line that adds the polyfill. If so, fixing the bug is a piece of cake.

**Comment by DaanPepping** [comment]
i also tried with version 20.2.0 without `forceUsingLegacyES5` but this results in the Promise.withResolvers is undefined error showing up again. when setting `forceUsingLegacyES5` to `true` the error disappears and the pdfviewer works as intended.

i did some digging in commit d26ba13 which i think is very close to version 20.2.0.

While I'm not to familiar with the source code but the `forceUsingLegacyES5` boolean seems to be used in `loadViewer()`: 
![image](https://github.com/user-attachments/assets/40eb0f38-3268-4b38-b29c-dd99f31d30d4)

NeedsES5 will resolve to true which caused the `GetPdfJsPath('viewer', true)` method to be called:
![image](https://github.com/user-attachments/assets/8a74d0fd-e454-4b6c-aeaa-5678b8c0e67a)

If i read the method correctly the resulting path becomes `assets/viewer-4.1.889-es5.js` 
when not using legacyES5 the path becomes `assets/viewer-4.1.889.js` 
weirdly enough i only see a `viewer-4.1.889.mjs` file in the assets folder so i'm not sure what is happening here.

i checked the [documentation](https://github.com/mozilla/pdf.js/wiki/Frequently-Asked-Questions#legacy-build) from pdf.js's latest release and there legacy build (previously called es5) should still support safari 16.4 and therefore not use Promise.withResolvers.

So i think if there would be a way to reintroduce the `forceUsingLegacyES5` property, i'd be able to target the legacy pdf.js viewer. 

Is there any way i can contribute to a fix?

**Comment by DaanPepping** [comment]
> Let's start with your hack. It's sort of OK to use it because I simply forgot to declare Angular 18 compatibility. You can make you hack a bit less hacky by adding some lines to the `package.json`. I had to do this in the showcase because another library also forgot to update to Angular 18:
> 
> <img alt="image" width="761" src="https://private-user-images.githubusercontent.com/3045767/362421628-dbaa2bc0-3c3d-4b63-afcc-30a988a8da85.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5MzQ4OTEsIm5iZiI6MTcyNDkzNDU5MSwicGF0aCI6Ii8zMDQ1NzY3LzM2MjQyMTYyOC1kYmFhMmJjMC0zYzNkLTRiNjMtYWZjYy0zMGE5ODhhOGRhODUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjlUMTIyOTUxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTY3ZGNmMmNjY2YyYWIzNzcxMWNkNTU4OWEzMmZhOWQzNTQ0ODkzMzZlZjMwMzMzMWY2ZDljOTAzMDdhYTk2NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Gg28jyY3ypw4q6lCtKkGxWQVqll3CBlIx-MyfLsouQc">

I tried this fix, but without changing the `package.lock.json` running `npm install` causes the following errors:
![image](https://github.com/user-attachments/assets/ffa3b8b4-83eb-496f-9b7d-00b4078785cd)

**Comment by stephanrauh** [comment]
We're getting closer (although I'm still confused).  Which version of zone.js are you using? Several versions of zone.js messed up with `Promise.withResolvers` on browsers that don't implement `Promise.withResolvers`. Zone.js wraps a non-existing function, resulting in a `TypeError`.

You can add the polyfill to the `index.html` or - if you've project before Angular 18 - to the `polyfills.ts` file. The bottom line is that the polyfill is loaded before zone.js is loaded. Here's the current documentation on polyfills: https://angular.dev/reference/versions#polyfills-for-non-cli-users

And here's the polyfill you need. You've already seen it, but let me repeat it for the sake of convenience to future readers:

```ts
    const TypelessPromise = Promise as any;
    if (!TypelessPromise.withResolvers) {
      TypelessPromise.withResolvers = function withResolvers() {
        let a: unknown;
        let b: unknown;
        const c = new this(function (resolve: unknown, reject: unknown) {
          a = resolve;
          b = reject;
        });
        return { resolve: a, reject: b, promise: c };
      };
    }
```

**Comment by stephanrauh** [comment]
Remains the question why your application detects it needs the ES5 version but fails to load it. You say you don't have the *-es5.mjs files in your assets folder. That's strange. I've dropped the minified version, but the non-minified version should still be there:
https://unpkg.com/browse/ngx-extended-pdf-viewer@21.3.6/assets/
<img width="961" alt="image" src="https://github.com/user-attachments/assets/3dcbf16b-56bc-4254-9781-97de3d1e11aa">

**Comment by stephanrauh** [comment]
If you want to add the lost attribute `forceUsingLegacyES5` again, you're welcome. Here's what to do:

- Add the `@Input() forceUsingLegacyES5` to `NgxExtendedPdfViewerComponent` again
- Add the parameter `forceUsingLegacyES5` to  `public async ensurePdfJsHasBeenLoaded(useInlineScripts: boolean)`
- use the parameter in this method
- and fix the missing parameter when the method is called.

I'm looking forward to your pull request!

**Comment by DaanPepping** [comment]
> We're getting closer (although I'm still confused). Which version of zone.js are you using? Several versions of zone.js messed up with `Promise.withResolvers` on browsers that don't implement `Promise.withResolvers`. Zone.js wraps a non-existing function, resulting in a `TypeError`.
> 
> You can add the polyfill to the `index.html` or - if you've project before Angular 18 - to the `polyfills.ts` file. The bottom line is that the polyfill is loaded before zone.js is loaded. Here's the current documentation on polyfills: https://angular.dev/reference/versions#polyfills-for-non-cli-users

I checked the documentation you linked in case i missed something but i think i'm doing things correctly. I'm using zone.js

**Comment by DaanPepping** [comment]
i'll try to get around creating a pr sometime this week for the `forceUsingLegacyES5`property :D

**Comment by stephanrauh** [comment]
Sounds good. More to the point, your error analysis sounds frustrating, but offering a PR sounds good.

While you're at it: I suspect the gulp file of https://github.com/stephanrauh/pdf.js doesn't transpile `Promise.withResolvers`. Maybe you can check this. To do so, clone the repository, run `npm install`, followed by `gulp generic-legacy`. This command builds the ES5 files somewhere in the dist folder. Open the files and check if there's a `Promise.withResolvers`. If so, can I talk you into finding out how to configure gulp correctly?

**Comment by alessandrocolinucci** [comment]
@stephanrauh @DaanPepping any news on this one?

**Comment by stephanrauh** [comment]
No news from me. @DaanPepping Are you still planning to submit a pull request, or do you want me to have a look at it?

**Comment by MathieuSommit** [comment]
I have seen that a polyfill already exists in `pdf-script-loader.service.ts:289`

But after testing Safari create a isolation for workers, this isolation can cause issues when trying to use polyfills within a worker, as they need to be explicitly included or polyfilled within the worker script itself.

A tmp fix after the build of our app, we inject this code at the end of `pdf.worker-4.6.687.min.mjs`
```
// Polyfills for Safari <17.4
if (!Promise.withResolvers) {
    Promise.withResolvers = function withResolvers() {
        let a;
        let b;
        const c = new this(function (resolve, reject) {
            a = resolve;
            b = reject;
        });
        return {resolve: a, reject: b, promise: c};
    };
}
```

**Comment by kasc-fidentity** [comment]
@stephanrauh We are experiencing the same issue using Angular 18. With iOS < 17.4, the PDF fails to render because of Promise.withResolvers.

**Comment by AdrianM** [comment]
+1

**Comment by kasc-fidentity** [comment]
We were able to work around it by overriding the following properties and functions in our component's constructor until the underlying problem is fixed in ngx-extended-pdf-viewer for iOS < 17.4

```
    const customPdfScriptLoaderService = this.pdfScriptLoaderService as any;
    customPdfScriptLoaderService.forceUsingLegacyES5 = true;
    customPdfScriptLoaderService._needsES5 = true;
    customPdfScriptLoaderService.needsES5 = (_: boolean) => {
        return true;
    };
    pdfDefaultOptions.needsES5 = true;
```

**Comment by devcomponent** [comment]
just setting 

pdfDefaultOptions.needsES5 = true;

was enough for me. 
thx

**Comment by stephanrauh** [comment]
Version 21.4.6 re-introduces the attribute `[forceUsingLegacyES5]`.

**Comment by Shivamchauhan12** [comment]
i have added polyfills externally (changed  the version of pdf viewer from 22 to 23 latest and the versions between them still not working in safari 16 ,15 or below

**Comment by Shivam3062** [comment]
i have added polyfills externally (changed the version of pdf viewer from 22 to 23 latest and the versions between them still not working in safari 16 ,15 or below

**Comment by stephanrauh** [comment]
@Shivam3062 Probably it's a good idea to hunt down the issue yourself. I'd like to help, but at the moment, I'm short of time, and I don't have old version of Safari, Chrome or Firefox.

Have a look at the class `PDFScriptLoaderService` and add breakpoints. For example, this class already polyfills the missing function under certain circumstances:

```ts
  /**
   * Angular 16 uses zone.js 0.13.3, and this version has a problem with Promise.withResolvers.
   * If your browser supports Promise.withResolvers, zone.js accidentally overwrites it with "undefined".
   * This method adds a polyfill for Promise.withResolvers if it is not available.
   */
  private polyfillPromiseWithResolversIfZoneJSOverwritesIt() {
    const TypelessPromise = Promise as any;
    if (!TypelessPromise.withResolvers) {
      TypelessPromise.withResolvers = function withResolvers() {
        let a: unknown;
        let b: unknown;
        const c = new this(function (resolve: unknown, reject: unknown) {
          a = resolve;
          b = reject;
        });
        return { resolve: a, reject: b, promise: c };
      };
    }
  }
```

If you find a bug, please tell me, so I can fix it and everybody benefits from your effort. Thanks in advance!

Best regards,
Stephan

**Comment by Shivam3062** [comment]
I will try my best to get the root cause of this  issue.
________________________________

**Comment by stephanrauh** [comment]
@AndresPuglia98 Do you really need to use the pdf.worker.min.mjs file from pdf.js? It should work, but it's always risky to mix the original pdf.js with my fork. However, I didn't change much in the worker file, so you'll probably get away with it. :)

**Comment by AndresPuglia98** [comment]
In the end I just used this and it works fine:
```ts
// @ts-expect-error This does not exist outside of polyfill which this is doing
if (typeof Promise.withResolvers === 'undefined') {
    if (window)
        // @ts-expect-error This does not exist outside of polyfill which this is doing
        window.Promise.withResolvers = function () {
            let resolve, reject;
            const promise = new Promise((res, rej) => {
                resolve = res;
                reject = rej;
            });
            return { promise, resolve, reject };
        };
}
```

**Comment by stephanrauh** [comment]
Where did you add this code? In the constructor of the component using the PDF viewer?

**Comment by AndresPuglia98** [comment]
I don't actually use `ngx-extended-pdf-viewer`, just came across this issue.
In my project I use it before setting the `pdfjs.GlobalWorkerOptions.workerSrc` param.

**Comment by stephanrauh** [comment]
I see. Thanks a lot for chiming in! 

I'm a bit confused because my source code contains pretty much the code you've posted, but people keep telling me it doesn't work. So I believe your code snippet may help. But it might be a challenge to find the correct event.

**Comment by Shivam3062** [comment]
![Image](https://github.com/user-attachments/assets/943bfa43-d43c-4786-aaf9-6e5cd659f99f)

**Comment by stephanrauh** [comment]
Yes, it's possible that you're trying to run the PDF viewer on a browser that's too old. Safari 14 is only five years old, and it's a shame to abandon devices that are five years old. On the other hand, Apple usually allows users to get updates for at least four years, so we're talking about devices that are nine years old. Or users reluctant to update their devices, but that's nothing I want to support because of IT security reasons.

In other words: I'd like to support Safari 14. But maybe that's impossible. The base library, pdf.js, is specifically written for an evergreen browser. Their target platform is the latest version of Firefox. If it runs elsewhere, that's nice, but it's no core issue to the pdf.js team.

I'm improving compatibility a few years by adding the "legacy" build, originally targeted at IE11. That's the *ES5 files, and I guess that's also the `.browserslistrc` file you've found. Or maybe not. The list that's really important is here: https://github.com/stephanrauh/pdf.js/blob/bleeding-edge/gulpfile.mjs#L81-L92. Note to myself: I ought to align both lists.

Safari, in particular, is always a bit difficult. I don't know what's wrong with Safari, but it's always been riddled by unexpected compatibility issues.

If supporting old browsers is an issue to you, try using older versions of ngx-extended-pdf-viewer. Just be aware that version 19.x is vulnerable to https://github.com/advisories/GHSA-wgrm-67xf-hhpq. As far as I remember, the vulnerability started roughly at version 17. Apart from that, going back to an old version means you're loosing some features, you're loosing my support (because my spare time is limited), and you might catch a few security issues. If you're ready to live with the tradeoffs, give it a try!

**Comment by stephanrauh** [comment]
Oh, wait. The message "unexpected keyword 'export'" puzzles me. The worker file accidentally has an export statement. That's weird.

**Comment by stephanrauh** [comment]
I've prepared the bug fix without publishing it yet. It'll be part of version 23.0.0-alpha.9 soon.

**Comment by Shivamchauhan12** [comment]
Ok thanks a lot but can you give a date when you thinking to be published

**Comment by stephanrauh** [comment]
Tomorrow or Monday. As usual, no promises, but that's my plan.

**Comment by Shivam3062** [comment]
![Image](https://github.com/user-attachments/assets/8b9bbca1-f302-48b0-8a9b-68b942fc41cd)

after upadting pdf-viewer to 23.0.0 i am exposed to new issue which is in the screenshot i am  little bit confused

**Comment by luca-peruzzo** [comment]
> ![Image](https://github.com/user-attachments/assets/8b9bbca1-f302-48b0-8a9b-68b942fc41cd)
> 
> ![Image](https://github.com/user-attachments/assets/99f1c787-d2a2-420b-9ca4-748b4c2a2d6c)
> 
> after updating pdf-viewer to 23.0.0 I am exposed to new issue which is in the screenshot when i am opening pdf in safari 15,16 first image issue occur in safari 15 and second one occured in safari 16

I have the same error on chrome, using `forceUsingLegacyES5` (I'm using "type": "module" on my package.json):
```
pdf.worker-4.10.712-es5.mjs:5784 Uncaught ReferenceError: exports is not defined
    at pdf.worker-4.10.712-es5.mjs:5784:33
    at pdf.worker-4.10.712-es5.mjs:5787:12
(anonymous) @ pdf.worker-4.10.712-es5.mjs:5784
(anonymous) @ pdf.worker-4.10.712-es5.mjs:5787
```

**Comment by Shivam3062** [comment]
any response @stephanrauh  please can you resolve this issue at your earliest

**Comment by stephanrauh** [comment]
The issue is tracked here: #2868

**Comment by Shivam3062** [comment]
As I described earlier when I open pdf in safari 15 and mac 11 only starting four pages content is not rendered correctly  all pages and thumbnails are blank after  initial 3 or 4 pages .

**Comment by Shivam3062** [comment]
HI @stephanrauh  IS 

> Hi, [@stephanrauh](https://github.com/stephanrauh) I am using 23.1.0 of pdf-viewer As I described earlier when I open pdf in safari 15 and mac 11 only starting four pages content is rendered correctly otherwise all the other pages and thumbnails are blank after initial 3 or 4 pages .
> 
> this issue i still there
> 
> ![Image](https://github.com/user-attachments/assets/b8eaf36e-3bf3-4301-8cf7-62db57244d32)

Is there anything that you can do to resolve this bug

**Comment by stephanrauh** [comment]
As long as I can't reproduce the bug, there's nothing I can do.

**Comment by stephanrauh** [comment]
Researching the issue with ChatGPT reveals that this is an issue with Safari 15. It seems they've fixed it in Safari 15.4 and above. In theory, I could implement a workaround - but this is probably only the beginning of a long list of incompatibilities. I'm afraid the library is not compatible to Safari 15.

By the way, this is an open-source project, and I'm running it in my leisure time. That, in turn, is limited, as you've noticed. You could help me by trying to hunt down the problem yourself. The "exports is not defined" bug was my task, but this follow-up issue was an easy investigation.

Thanks in advance,
Stephan

**Comment by stephanrauh** [comment]
I've found a configuration error. Now (when I publish version 23.1.1), compatibility is a little better. For some reason, it was messed up even for Safari 17. Now the minimum requirements seems to be iOS 16 or 17 and Safari 16. I'm not entirely sure because Browserstack only lists the version number of the OS, but not the version number of Safari. On MacOS, it's a bit more talkative: the showcase runs on Safari 16.5 and above.

In older browsers / iOS versions the showcase doesn't start, but that's possibly a problem of the showcase and not a problem of the library. However, as mentioned above, I don't believe Safari 15 is supported. I've configured Babel to emit code compatible to Safari 15, but even Babel can't solve every compatibility issue.
